packages/medusa/src/strategies/product/import.ts

```diff
@@ -0,0 +1,772 @@
+                (createdProductOption) =>
+                  createdProductOption.title === variantOption.title
+              )!.id
+          ) || []
+
+        variant.options =
+          (variant.options as object[])?.map((o, index) => ({
+            ...o,
+            option_id: optionIds[index],
+          })) || []
+
+        await this.productVariantService_
+          .withTransaction(transactionManager)
+          .create(product!, variant as unknown as CreateProductVariantInput)
+
+        this.updateProgress_(batchJobId)
+      } catch (e) {
+        ProductImportStrategy.throwDescriptiveError(variantOp)
+      }
+    }
+  }
+
+  /**
+   * Method updates product variants from a CSV data.
+   *
+   * @param batchJobId - An id of the current batch job being processed.
+   * @param transactionManager - Transaction manager responsible for current batch import.
+   */
+  private async updateVariants(
+    batchJobId: string,
+    transactionManager: EntityManager
+  ): Promise<void> {
+    const productOptionRepo = this.manager_.getCustomRepository(
+      this.productOptionRepo_
+    )
+
+    const variantOps = await this.getImportDataFromRedis(
+      batchJobId,
+      OperationType.VariantUpdate
+    )
+
+    for (const variantOp of variantOps) {
+      try {
+        await this.prepareVariantOptions(variantOp, productOptionRepo)
+
+        await this.productVariantService_
+          .withTransaction(transactionManager)
+          .update(
+            variantOp["variant.id"] as string,
+            transformVariantData(variantOp) as UpdateProductVariantInput
+          )
+      } catch (e) {
+        ProductImportStrategy.throwDescriptiveError(variantOp)
+      }
+
+      this.updateProgress_(batchJobId)
+    }
+  }
+
+  /**
+   * Extend records used for creating variant options with corresponding product option ids.
+   *
+   * @param variantOp - Parsed row data form CSV
+   * @param productOptionRepo - ProductOption repository
+   */
+  protected async prepareVariantOptions(
+    variantOp,
+    productOptionRepo: ProductOptionRepository
+  ): Promise<void> {
+    const productOptions = variantOp["variant.options"] || []
+
+    for (const o of productOptions) {
+      const { id } = (await productOptionRepo.findOne({
+        where: { title: o._title },
+      })) as ProductOption
+
+      o.option_id = id
+    }
+  }
+
+  /**
+   * Store import ops JSON to Redis.
+   * Data will expire after an hour.
+   *
+   * @param batchJobId - An id of the current batch job being processed.
+   * @param results - An object containing parsed CSV data.
+   */
+  async setImportDataToRedis(
+    batchJobId: string,
+    results: Record<OperationType, TParsedRowData[]>
+  ): Promise<void> {
+    for (const op in results) {
+      if (results[op]?.length) {
+        await this.redisClient_.set(
+          `pij_${batchJobId}:${op}`,
+          JSON.stringify(results[op]),
+          "EX",
+          60 * 60
+        )
```

Comment:
thought(non-blocking): I have concerns about how much memory this takes up given an import of 100, 1 000, 10 000 rows.

My initial thought was actually to save the "instructions" in a temporary file that is uploaded via the `fileService`.

Some reasons for this:
- generally "cold storage" is cheap so we can store almost unlimited amounts of data (it is ofc slower than Redis; however, speed is not super important in imports). In comparison, a Redis instance with 250MB is fairly expensive (e.g. on Heroku it costs ~$60/month).
- atm we primarily use Redis for sub/pub in the event bus; this usage means that we don't have a huge dependency on Redis; for example, we could easily imagine replacing Redis for another message queue technology. If we use Redis for storage, however, we create a larger dependency on Redis. This could become a restriction in the future when considering ways of deploying Medusa.

This is more of a consideration that I think we will have to address in the future than it is a change request; however, if you have time to look into what it would take to change the storage of instructions to be in a temporary file instead I think that would be helpful :)

Reply:
I went with Redis since it was the initial plan and also since we agreed that we will limit the size of a file to 15mb (this needs to be implemented yet), however, these are really good points, especially the second one architecture wise - will make the change :)

Reply:
Nice!